Can't build it for AUR packaging

Hi,

I'm trying to package mdns-scanner for Archlinux AUR repository, and I can't build it in a clean chroot environment. I believe some configuration or dependency is missing, although the README says it doesn't require any dependency in Linux other than the Rust toolchain.

The error log is very extensive, but I could not find anything that could help me. I'm going to upload it here, for you. Any help would be appreciated.

[mdns-scanner-0.12.1-1-x86_64-build.log](https://github.com/user-attachments/files/20826695/mdns-scanner-0.12.1-1-x86_64-build.log)

Thanks,

Kemel Zaidan

Hi Kemel,

It seems like you're building for `x86_64-unknown-linux-gnu`, have you tried using the `x86_64-unknown-linux-musl` target instead? 

I purposefully only build for musl to ensure all binaries are statically linked and thus avoiding runtime dependencies on even GLIBC (except for windows where Npcap can't be statically linked).

The build output suggests that [jemalloc](https://crates.io/crates/tikv-jemallocator) is not available at compile time.

```
error: linking with `cc` failed: exit status: 1
  |
  = note:  "cc" "-m64" "/tmp/rustcCGE91O/symbols.o" "<1 object files omitted>" "-Wl,--as-needed" "-Wl,-Bstatic" "/tmp/rustcCGE91O/{libtikv_jemalloc_sys-f289774c825282fb.rlib,libring-6997716407ff100a.rlib}.rlib" "<sysroot>/lib64/rustlib/x86_64-unknown-linux-gnu/lib/{libcompiler_builtins-*}.rlib" "-Wl,-Bdynamic" "-lgcc_s" "-lutil" "-lrt" "-lpthread" "-lm" "-ldl" "-lc" "-L" "/tmp/rustcCGE91O/raw-dylibs" "-Wl,--eh-frame-hdr" "-Wl,-z,noexecstack" "-L" "/build/mdns-scanner/src/mdns-scanner-0.12.1/target/release/build/ring-9dafb2210497415a/out" "-L" "/build/mdns-scanner/src/mdns-scanner-0.12.1/target/release/build/tikv-jemalloc-sys-3179bfc0819d7c7a/out/build/lib" "-L" "<sysroot>/lib64/rustlib/x86_64-unknown-linux-gnu/lib" "-o" "/build/mdns-scanner/src/mdns-scanner-0.12.1/target/release/deps/mdns_scanner-10883353f01b20f9" "-Wl,--gc-sections" "-pie" "-Wl,-z,relro,-z,now" "-Wl,-O1" "-Wl,--strip-all" "-nodefaultlibs"
  = note: some arguments are omitted. use `--verbose` to show all linker arguments
  = note: /usr/bin/ld: /build/mdns-scanner/src/mdns-scanner-0.12.1/target/release/deps/mdns_scanner-10883353f01b20f9.mdns_scanner.5808370a1dee65cf-cgu.0.rcgu.o: in function `<tikv_jemallocator::Jemalloc as core::alloc::global::GlobalAlloc>::dealloc':
          /build/.cargo/registry/src/index.crates.io-1949cf8c6b5b557f/tikv-jemallocator-0.6.0/src/lib.rs:129:(.text._ZN10axoprocess3Cmd3env17h2ab18a2f74f2d946E+0x43): undefined reference to `_rjem_sdallocx'
```

EDIT: You can try to commenting out this part if you want to see if you can compile it with the standard allocator, but targeting musl would be preferable

https://github.com/CramBL/mdns-scanner/blob/ff33462769e73dded147223f23a298e350111d36/src/main.rs#L16

If you can provide more information about your build setup and/or constraints then I will add it to CI if it makes sense so that it's supported in the future

--- a/.github/workflows/build-binaries.yml
+++ b/.github/workflows/build-binaries.yml
@@ -46,7 +46,7 @@ jobs:
 
         - name: "Build ${{ matrix.target }}"
           run: |
-            cargo build --release --locked --target ${{ matrix.target }} --features self-update
+            cargo build --release --locked --target ${{ matrix.target }} --features dist
 
         - name: "Test binary"
           if: ${{ matrix.target == 'aarch64-apple-darwin' }}
@@ -92,7 +92,7 @@ jobs:
           targets: ${{ matrix.target }}
 
       - name: Build ${{ matrix.target }}
-        run: cargo build --release --locked --target ${{ matrix.target }} --features self-update
+        run: cargo build --release --locked --target ${{ matrix.target }} --features dist
 
       - name: "Archive ${{ matrix.target }}"
         shell: bash
@@ -134,7 +134,7 @@ jobs:
           fi
       - name: "Build binary"
         run: |
-          cargo zigbuild --release --locked --target ${{ matrix.platform.target }}.2.17 --features self-update
+          cargo zigbuild --release --locked --target ${{ matrix.platform.target }}.2.17 --features dist
         env:
           CARGO_TARGET_S390X_UNKNOWN_LINUX_GNU_LINKER: s390x-linux-gnu-gcc
       - uses: uraimo/run-on-arch-action@aa8e672a03e10b403401927ea2ceda57e1b68ac3 # v3
@@ -184,7 +184,7 @@ jobs:
           rustup target add ${{ matrix.platform.target }}
       - name: "Build binary"
         run: |
-          cargo build --release --locked --target ${{ matrix.platform.target }} --features self-update
+          cargo build --release --locked --target ${{ matrix.platform.target }} --features dist
         env:
           CARGO_TARGET_POWERPC64LE_UNKNOWN_LINUX_GNU_LINKER: powerpc64le-linux-gnu-gcc
           CARGO_TARGET_POWERPC64_UNKNOWN_LINUX_GNU_LINKER: powerpc64-linux-gnu-gcc
@@ -232,7 +232,7 @@ jobs:
             pip3 install cargo-zigbuild
           fi
       - name: "Build binary"
-        run: cargo zigbuild --release --locked --target ${{ matrix.platform.target }} --features self-update
+        run: cargo zigbuild --release --locked --target ${{ matrix.platform.target }} --features dist
       - uses: uraimo/run-on-arch-action@aa8e672a03e10b403401927ea2ceda57e1b68ac3 # v3
         name: "Test binary"
         with:
@@ -283,7 +283,7 @@ jobs:
             pip3 install cargo-zigbuild
           fi
       - name: "Build binary"
-        run: cargo zigbuild --release --locked --target ${{ matrix.target }} --features self-update
+        run: cargo zigbuild --release --locked --target ${{ matrix.target }} --features dist
 
       - name: "Test binary"
         if: matrix.target == 'x86_64-unknown-linux-musl'
@@ -331,7 +331,7 @@ jobs:
           rustup target add ${{ matrix.platform.target }}
       - name: "Build binary"
         run: |
-          cargo build --release --locked --target ${{ matrix.platform.target }} --features self-update
+          cargo build --release --locked --target ${{ matrix.platform.target }} --features dist
         env:
           CARGO_TARGET_AARCH64_UNKNOWN_LINUX_MUSL_LINKER: aarch64-linux-gnu-gcc
           CARGO_TARGET_ARMV7_UNKNOWN_LINUX_MUSLEABIHF_LINKER: arm-linux-gnueabihf-gcc

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -9,6 +9,7 @@
 ### Changed
 
 - Significantly reduce binary size by scrutinizing dependencies
+- To simplify installation from source with the default profile, `jemalloc` is feature-gated. While it's included in prebuilt binaries, users installing from source must explicitly enable the `jemalloc` feature to use it instead of the system allocator, i.e.: `cargo install --git https://github.com/CramBL/mdns-scanner mdns-scanner --features jemalloc`
 
 ## [0.12.1] - 2025-06-15
 

--- a/Cargo.toml
+++ b/Cargo.toml
@@ -27,7 +27,7 @@ axoupdater = { workspace = true, features = [
 mimalloc = "0.1.47"
 
 [target.'cfg(all(not(target_os = "windows"), not(target_os = "openbsd"), not(target_os = "freebsd"), any(target_arch = "x86_64", target_arch = "aarch64", target_arch = "powerpc64")))'.dependencies]
-tikv-jemallocator = "0.6.0"
+tikv-jemallocator = { version = "0.6.0", optional = true }
 
 [dev-dependencies]
 assert_cmd = "2.0.17"
@@ -49,6 +49,9 @@ inherits = "release"
 
 [features]
 default = []
+# Toggle for the features that are enabled for the distributed prebuilt version
+dist = ["jemalloc", "self-update"]
+jemalloc = ["tikv-jemallocator"]
 # Adds self-update functionality. This feature is only enabled for the cargo-dist installer
 # and should be left unselected when building mdns-scanner for package managers.
 self-update = ["axoupdater", "mds-cli/self-update"]

--- a/dist-workspace.toml
+++ b/dist-workspace.toml
@@ -20,7 +20,7 @@ install-path = ["$XDG_BIN_HOME/", "$XDG_DATA_HOME/../bin", "~/.local/bin"]
 # Whether to install an updater program
 install-updater = false
 # Which actions to run on pull requests
-pr-run-mode = "upload"
+pr-run-mode = "plan"
 # Build only the required packages, and individually
 precise-builds = true
 # Whether CI should include auto-generated code to build local artifacts

--- a/src/main.rs
+++ b/src/main.rs
@@ -3,6 +3,7 @@
 static GLOBAL: mimalloc::MiMalloc = mimalloc::MiMalloc;
 
 #[cfg(all(
+    feature = "jemalloc",
     not(target_os = "windows"),
     not(target_os = "openbsd"),
     not(target_os = "freebsd"),

